File: backend/auth/middleware.py
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.requests import Request
from starlette.responses import Response, JSONResponse
from typing import Callable, Awaitable
from ..auth import firebase_auth
import firebase_admin

class FirebaseAuthMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        # Clear any previous user state
        request.state.user = None

        auth_header = request.headers.get("Authorization")
        if not auth_header:
            # If no token, just proceed. The route dependency will handle protection.
            return await call_next(request)

        try:
            scheme, token = auth_header.split()
            if scheme.lower() == "bearer":
                # Verify the token
                decoded_token = firebase_auth.verify_id_token(token)
                # Attach the decoded token to the request state for use in endpoints
                request.state.user = decoded_token
        except (ValueError, firebase_admin.auth.InvalidIdTokenError) as e:
            # This will catch malformed headers or invalid tokens.
            # We can return an error immediately for invalid tokens.
            return JSONResponse(
                status_code=401,
                content={"detail": f"Invalid authentication token: {e}"},
            )
        except Exception as e:
            # Catch other potential verification errors
            return JSONResponse(
                status_code=500,
                content={"detail": f"Could not process authentication token: {e}"},
            )

        response = await call_next(request)
        return response

File: backend/auth/middleware.js
/**
 * Firebase Authentication Middleware
 */
const { verifyIdToken } = require('./firebaseAuth');

/**
 * Middleware that extracts and verifies Firebase ID token from Authorization header.
 * Attaches the decoded user to req.user if valid.
 */
async function firebaseAuthMiddleware(req, res, next) {
    // Clear any previous user state
    req.user = null;

    const authHeader = req.headers.authorization;

    if (!authHeader) {
        // If no token, just proceed. Route-level protection will handle it.
        return next();
    }

    try {
        const parts = authHeader.split(' ');
        if (parts.length !== 2 || parts[0].toLowerCase() !== 'bearer') {
            return next(); // Invalid format, let route handle protection
        }

        const token = parts[1];
        const decodedToken = await verifyIdToken(token);

        // Attach the decoded token to the request for use in endpoints
        req.user = decodedToken;
        next();
    } catch (error) {
        // For invalid tokens, return error immediately
        if (error.status === 401) {
            return res.status(401).json({
                detail: `Invalid authentication token: ${error.message}`
            });
        }
        // For other errors, return 500
        return res.status(500).json({
            detail: `Could not process authentication token: ${error.message}`
        });
    }
}

module.exports = {
    firebaseAuthMiddleware
};
